build-tools/packages/build-tools/src/fluidBuild/tasks/leaf/biomeTasks.ts

```diff
@@ -0,0 +1,74 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "path";
+import { GitRepo } from "../../../common/gitRepo";
+import { LeafWithFileStatDoneFileTask } from "./leafTask";
+import { getResolvedFluidRoot } from "../../../common/fluidUtils";
+import findUp from "find-up";
+
+export class BiomeTask extends LeafWithFileStatDoneFileTask {
+	private repoRoot: string | undefined;
+	private gitRepo: GitRepo | undefined;
+
+	/**
+	 * Includes all files in the task's package directory and any biome config files in the directory tree. Files ignored
+	 * by git are excluded.
+	 */
+	protected async getInputFiles(): Promise<string[]> {
+		this.repoRoot ??= await getResolvedFluidRoot(true);
+		this.gitRepo ??= new GitRepo(this.repoRoot);
```

Comment:
I was going to suggest lazily initialized properties, but I guess in the case of 'this.repoRoot' the current pattern is a micro-optimization that skips the unnecessary await.  Cute pattern.

Reply:
I think I come down on the side of less cute pattern. This is really what promises are for. This pattern leads to the lint warnings about non-null assertions, for example, as few lines below.

Reply:
Maybe I am missing something, but repoRoot can only be retrieved asynchronously (`getResolvedFluidRoot` is async), so the value can't be used in a Task constructor. I need the concrete value, not a Promise. Passing a promise around does not seem any clearer, but maybe that's because I can't stand the non-await Promise code. It's verbose and difficult to reason about.

I don't see how to avoid repoRoot possibly being undefined if it can't be set in the constructor. An async creation function also doesn't help as it is, because that would require much more churn to the task infra to use that function instead of the contstructor.

Reply:
What you really need is a Promise. Promises are efficient and manage that value or undefined state you are working with for you.
Here is an example of getting started: https://github.com/jason-ha/FluidFramework/commit/9f3d7625c41294d69252d6ead3899a2b5dbab7d8

Reply:
Ahhh, it's that pesky annoying .then stuff. Ok I'll open another pr and fix this.

Reply:
Fixed in https://github.com/microsoft/FluidFramework/pull/20275

Reply:
You don't have to use `.then`, but there are a number of cases where it is much easier to read than the `await` version, which is really just syntactic sugar. The new GitRepo case is one.